### ISSUE TITLE
Checks ensuring curated TSVs/CSVs have consistent line endings

### ISSUE THREAD

**Issue by kaiiam** [INCATools/ontology-development-kit#375]
Referencing https://github.com/EnvironmentOntology/envo/issues/1026, ontologies maintaining curated TSV/CSV files, e.g. inputs to DOSDPs or ROBOT templates, suffer from the issue that line endings are different on PCs vs UNIX/LINUX systems. This creates an issue with overly complicated (git) diffs because the change between line endings will rewrite all rows in the tabular file, instead of just the intended changes, making it more difficult to review pull requests. 

In ENVO @cmungall is suggesting the use of travis checks to ensure that tsv/csv files are maintained as UNIX, but it great if something like that could be built into ODK or standardized such that it can fix this problem across the board for any ODK and or OBO ontology. Better to fix this problem once then have everyone have to do it on their own with different solutions, e.g. OBI has some custom code to maintain their Robot template csvs with PC line endings.

**Comment by jamesaoverton** [comment]
This is the OBI script: https://github.com/obi-ontology/obi/blob/master/src/scripts/sort-templates.py#L20

Basically we just run everything through Python `csv.writer(tsv, delimiter="\t", lineterminator="\n")` -- tabs and Unix line endings, with consistent quoting.

**Comment by matentzn** [comment]
yeah we should do something like that, great call. Thanks for the code @jamesaoverton

**Comment by kaiiam** [comment]
Much appreciated! Thanks for the quick responses!

**Comment by matentzn** [comment]
On second thought, this could be more complex than it sounds.. @kaiiam How do you see this fit into the general workflow? We cant really control, from the side of the ODK, how people save their tables.. Say someone opens a TSV on a windows machine with Excel and saves it, odds are it gets bad line endings. And I also don't see many people running a clean up command prior to committing a changed table manually. If the goal is better readable diffs, is this something we need to write Github actions for?

**Comment by kaiiam** [comment]
I'm not certain what the best method would be, @cmungall's suggestion for ENVO was to build in travis checks to ensure line ending consistency, but of course those type of build checks are downstream of ODK. I'm not familiar with github actions so I can't comment there. 

With @jamesaoverton's solution there is a "pre-commit hook" line linking to the line ending checking script see [obi#initial-set-up](https://github.com/obi-ontology/obi#initial-set-up) where users should run `ln -s ../../src/scripts/check-line-endings.sh .git/hooks/pre-commit` prior to commiting. Perhaps something like that could be a standard part of ODK's files, the python script plus the git hook in the makefile?

**Comment by matentzn** [comment]
regarding the second though, there is no workflow that requires anyone to run anything before committing.. Thats my main issue. Having a check for wrong line endings in travis is very easy - but yeah it would require the user to come back and fix them somehow, which 99% of our users wouldn't know how to do.. Maybe other people have an idea

**Comment by kaiiam** [comment]
Would incorporating the hook into the initial ODK release commit not have the changes stick? If you cloned the repo into a new location would the `.git/hook` not get cloned as well?

**Comment by matentzn** [comment]
Ah now I get it.. Ok. Maybe! I will consider doing this ticket, but in calmer waters. :) Thanks for the help @kaiiam

**Comment by kaiiam** [comment]
@nicolevasilevsky no worries it's probably not super high priority, but I think it'll be helpful later when ODK and robot are more commonly used together. I'll bug you about this again in the future. Thanks!

**Comment by nicolevasilevsky** [comment]
I think you meant to tag @matentzn, @kaiiam :)

**Comment by kaiiam** [comment]
Sorry @nicolevasilevsky yes I did.

**Comment by cmungall** [comment]
let's not close this ticket yet, at least use this as a locus for aligning how we do this across repos

> regarding the second though, there is no workflow that requires anyone to run anything before committing.. Thats my main issue. Having a check for wrong line endings in travis is very easy - but yeah it would require the user to come back and fix them somehow, which 99% of our users wouldn't know how to do

we just teach them? Trivial on max/linux. Sure there must be an easy solution for windows

We could use gh-actions to normalize, but we end up with awful spurious diffs. The general rule I always instill in people working on ontology repo is:

 - always always always do a git diff before commit (or whatever the equivalent in your client is)
 - if the diff is bigger than expected, do not commit. Seek help if you can't resolve yourself

**Comment by jamesaoverton** [comment]
In practise, Becky or I review all OBI PRs. We try to enforce clean diffs. If they're messy then we ask more technical submitters to run `make sort` themselves or we just do it ourselves. If they're really messy then we'll use a squash commit.

I just added an explicit `make sort` step to our README: https://github.com/obi-ontology/obi#making-changes

Since git hooks are in `.git/hooks` they aren't committed or shared with a repository. There's various solutions for sharing hooks but I haven't been satisfied with any that I've seen yet.

**Comment by matentzn** [comment]
I do not want, right now, to add the table check by default (i.e. adding it to the `test` target), because many tables out there are seriously broken. Instead, in the first instance, for the ODK 1.3.0 release, we will introduce a target for validating at the developers discretion, with some instructions on how to use this in CI. 

This is realised with the `tsvalid` tool @jamesaoverton and I have developed: https://github.com/ontodev/tsvalid

**Comment by matentzn** [comment]
Once the 1.3.0 is released, validation can be invoked like this:

```
sh run.sh make TSV=my.tsv validate-tsv 
```

or 

```
sh run.sh make validate-all-tsv
```

Which validates all TSV registered by ODK. This is, for the time being, all DOSDP template TSVs.

**Comment by matentzn** [comment]
cc @dosumis

**Comment by matentzn** [comment]
Let's open more specific checks for future table validation. Fixed in #530
